[dtensor][random] use torch.uint64 as the seed/offset tensor dtype to avoid overflow

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141532
* #141223
* #141220
* #141731

**Summary**
DTensor RNG code raises error if the seed passed in is beyong `torch.int64` range (e.g. `torch.tensor([2**64-1])` raises error). The solution is to specify the `dtype=torch.uint64` in the `torch.tensor()` call.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

XilunWu committed:
[dtensor][random] use torch.uint64 as the seed/offset tensor dtype to avoid overflow

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][random] use torch.uint64 as the seed/offset tensor dtype to avoid overflow"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][random] use torch.uint64 as the seed/offset tensor dtype to avoid overflow"


**Summary**
DTensor RNG code raises error if the seed passed in is beyong `torch.int64` range (e.g. `torch.tensor([2**64-1])` raises error). The solution is to specify the `dtype=torch.uint64` in the `torch.tensor()` call.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab reviewed (approved):
glad you found this, i noticed the error once but not consistently, and didn't track down where it came from.

Skylion007 commented on torch/distributed/tensor/_random.py:
```diff
@@ -198,7 +198,7 @@ def set_offset(self, name: str, offset: int) -> None:
 
         seed_tensor = (self.rng_states[name])[0:8]
-        offset_tensor = torch.tensor([offset]).view(torch.uint8)
+        offset_tensor = torch.tensor([offset], dtype=torch.uint64).view(torch.uint8)
```
Isn't this immediately cast to uint8?

wconstab replied:
offset comes in as a uint64 essentially.  we're using a tensor to convert it to its constituent bytes because this is the format used by torch's RNG state.  the initial dtype=uint64 is needed to prevent overflow warnings mainly.

Skylion007 reviewed (commented):
Bit confused here

Skylion007 commented:
Oh I see, this is just to hide an error.

XilunWu committed:
Update on "[dtensor][random] use torch.uint64 as the seed/offset tensor dtype to avoid overflow"


**Summary**
DTensor RNG code raises error if the seed passed in is beyong `torch.int64` range (e.g. `torch.tensor([2**64-1])` raises error). The solution is to specify the `dtype=torch.uint64` in the `torch.tensor()` call.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge -f "`python test/run_test.py --verbose -i test_optim -- -k test_forloop_goes_right_direction_multigpu` times out in CI but a local run attempt is successful. This failure should be unrelated to this PR."
